### PR TITLE
Add polls

### PR DESCRIPTION
**Pull request by sebm253**
- [x] https://github.com/discord/discord-api-docs/pull/6746

currently awaiting clarification on:
- [x] poll creation
- [ ] ~~updating polls in messages~~
- [ ] ~~editing messages with polls~~
- [x] poll builder?
- [x] new intent (instead of the reactions one)

### CODE REVIEW

File: events/listener_adapter.go [hunk 1]
```diff
@@ -147,6 +147,10 @@ type ListenerAdapter struct {
 	OnMessageUpdate func(event *MessageUpdate)
 	OnMessageDelete func(event *MessageDelete)
 
+	// Message Poll Events
```
Comment (mlnrDev) [inline events/listener_adapter.go:150]: do we want to split poll events into DM and guild events?

File: discord/poll.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/disgoorg/snowflake/v2"
+)
+
+type Poll struct {
+	Question         PollMedia      `json:"question"`
+	Answers          []PollAnswer   `json:"answers"`
+	Expiry           *time.Time     `json:"expiry"`
+	AllowMultiselect bool           `json:"allow_multiselect"`
+	LayoutType       PollLayoutType `json:"layout_type"`
+	Results          []PollResults  `json:"results"`
```
[the remark this answers is not in the document]
Comment (mlnrDev) [reply discord/poll.go]: ty

File: discord/message_create.go [hunk 3]
```diff
@@ -18,6 +18,7 @@ type MessageCreate struct {
 	MessageReference *MessageReference    `json:"message_reference,omitempty"`
 	Flags            MessageFlags         `json:"flags,omitempty"`
 	EnforceNonce     bool                 `json:"enforce_nonce,omitempty"`
+	Poll             PollCreate           `json:"poll,omitempty"`
```
Comment (topi314) [inline discord/message_create.go:21]: empty structs don't get omitted via `omitempty`, this should be a pointer

File: discord/message_create_builder.go [hunk 4]
```diff
@@ -251,6 +251,12 @@ func (b *MessageCreateBuilder) SetSuppressEmbeds(suppressEmbeds bool) *MessageCr
 	return b
 }
 
+// SetPoll sets the Poll of the Message
+func (b *MessageCreateBuilder) SetPoll(poll PollCreate) *MessageCreateBuilder {
```
Comment (topi314) [inline discord/message_create_builder.go:255]: we can accept either a pointer here or have a `ClearPoll` method

File: discord/webhook_message_create_builder.go [hunk 5]
```diff
@@ -210,6 +210,12 @@ func (b *WebhookMessageCreateBuilder) SetThreadName(threadName string) *WebhookM
 	return b
 }
 
+// SetPoll sets the Poll of the webhook Message
+func (b *WebhookMessageCreateBuilder) SetPoll(poll PollCreate) *WebhookMessageCreateBuilder {
```
Comment (topi314) [inline discord/webhook_message_create_builder.go:214]: we can accept either a pointer here or have a `ClearPoll` method

File: gateway/gateway_intents.go [hunk 6]
```diff
@@ -29,6 +29,10 @@ const (
 	_
 	IntentAutoModerationConfiguration
 	IntentAutoModerationExecution
+	_
+	_
+	IntentGuildMessagePolls
+	IntentDirectMessagePolls
```
Comment (topi314) [inline gateway/gateway_intents.go:35]: can we have a `IntentsMessagePolls`?
```go
IntentsMessagePolls = IntentGuildMessagePolls |
		IntentDirectMessagePolls
```

File: discord/webhook_message_create.go [hunk 7]
```diff
@@ -15,6 +15,7 @@ type WebhookMessageCreate struct {
 	Flags           MessageFlags         `json:"flags,omitempty"`
 	ThreadName      string               `json:"thread_name,omitempty"`
 	AppliedTags     []snowflake.ID       `json:"applied_tags,omitempty"`
+	Poll            PollCreate           `json:"poll,omitempty"`
```
Comment (topi314) [inline discord/webhook_message_create.go:18]: empty structs don't get omitted via `omitempty`, this should be a pointer

File: events/guild_message_poll_events.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package events
+
+import (
+	"github.com/disgoorg/disgo/discord"
+	"github.com/disgoorg/snowflake/v2"
+)
+
+// GenericGuildMessagePollVote is called upon receiving GuildMessagePollVoteAdd or GuildMessagePollVoteRemove (requires gateway.IntentGuildMessagePolls)
+type GenericGuildMessagePollVote struct {
```
Comment (topi314) [inline events/guild_message_poll_events.go:9]: do we want some more cache getters?
`Channel() (discord.GuildMessageChannel, bool)`
`Member() (discord.Member, bool)`

File: events/message_poll_events.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package events
+
+import (
+	"github.com/disgoorg/disgo/discord"
+	"github.com/disgoorg/snowflake/v2"
+)
+
+// GenericMessagePollVote is a generic poll vote event (requires gateway.IntentGuildMessagePolls or gateway.IntentDirectMessagePolls)
```
Comment (topi314) [inline events/message_poll_events.go:8]:
```suggestion
// GenericMessagePollVote is a generic poll vote event (requires gateway.IntentGuildMessagePolls and or gateway.IntentDirectMessagePolls)
```

File: events/message_poll_events.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+type GenericMessagePollVote struct {
```
Comment (topi314) [inline events/message_poll_events.go:9]: do we want some more cache getters?
`Channel() (discord.GuildMessageChannel, bool)`
`Member() (discord.Member, bool)`

**Review by topi314 — CHANGES_REQUESTED** [review]

File: events/message_poll_events.go (hunk 10 above)
Comment (mlnrDev) [reply events/message_poll_events.go:9]: channel sure, member probably not

File: events/guild_message_poll_events.go (hunk 8 above)
Comment (mlnrDev) [reply events/guild_message_poll_events.go:9]: channel sure, member probably not

File: discord/poll_create_builder.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+	return b
+}
+
+// ClearAnswers removes all answers of the Poll
+func (b *PollCreateBuilder) ClearAnswers() *PollCreateBuilder {
+	b.Answers = []PollMedia{}
+	return b
+}
+
+// SetDuration sets the duration of the Poll (in seconds)
```
[the remark this answers is not in the document]
Comment (mlnrDev) [reply discord/poll_create_builder.go:45]:
```suggestion
// SetDuration sets the duration of the Poll (in hours)
```

[the remark this answers is not in the document]
Comment (mlnrDev) [reply discord/poll_create_builder.go]: ty

**Review by topi314 — APPROVED** [review]

**Pull request merged**